### PR TITLE
Fix performance issue on Smarty template caching

### LINKED ISSUE #24235: Smarty template caching Module::isCached() faulty causes performance loss
#### Describe the bug

Prestashop class **Module** (`Prestashop /classes/module/Module.php`) has a function called **isCached** to check if tpl file is cached. This function fails to check and returns **false** while tpl is cached. 

Commit https://github.com/PrestaShop/PrestaShop/pull/13804

Person who made PR was fixing issue #9763 and modified Module's class function **fetch** adding this:

```
if ($compile_id === null) {
            $compile_id = Context::getContext()->shop->theme->getName();
        }
```
Because of this change when you call **fetch** function with **$compile_id = null** it generates smarty cache file under $compile_id directory which is theme name in this case (if $compile_id is null then $compile_id is set to theme name).

But when you try to check if template is cached with function **isCached** with **$compile_id = null** you always get **false** because code added in **fetch** function should be also added to **isCached** function.

Templates are fetched (cached) under compile_id which is theme name in most cases, but isCached function checks if template is cached with empty compile_id which causes false result.

**This bug causes a huge performance issues.** Many plugins check if tpl is cached and if not - they perform data aggregation and assignation for smarty to render tpl but actually is not needed. 

#### Steps to Reproduce

1. Turn on debug profiling;
2. Open Prestashop Front Office and check executed SQL queries count;
3. Add this code 

```
if ($compile_id === null) {
            $compile_id = Context::getContext()->shop->theme->getName();
        }
```

4. in the beggining of **Module** (`Prestashop /classes/module/Module.php`) class function **isCached**;
5. Open Prestashop Front Office in separate window and check SQL queries count;
6. Second time queries count should be lower;

#### Additional information

* PrestaShop version: From 1.7.7.0
* PHP version: N/A

### PR DESCRIPTION
**Pull request by lukaslau**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | From 1.7.7.0 version Module class function isCached fails to check if tpl is cached. This PR fixes that.
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #24235.
| How to test?      | In order to test if it is fixed you need to check if function isCached works properly or try to reproduce issue. How to reproduce issue is described in issue ticket.
| Possible impacts? | Smarty cache compile / fetch / clear

Cherry-pick PR for https://github.com/PrestaShop/PrestaShop/pull/24241

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25413)

### CODE REVIEW

File: classes/module/Module.php [hunk 1]
```diff
@@ -3515,6 +3528,16 @@ public function saveDashConfig(array $config)
     {
         return false;
     }
+
+    /**
+     * Returns shop theme name from context as a default compile Id.
+     *
+     * @return string
+     */
+    public function getDefaultCompileId(): string
```
Comment (PierreRambaud) [inline classes/module/Module.php]:
```suggestion
    public function getDefaultCompileId()
```

**Review by PierreRambaud — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]
lgtm...  I start the QA dev while CI is running

**Comment by matthieu-rolland** [comment]
### Without the PR, I empty smarty cache, then go on FO home page:

**First load:** 349 queries
**Second load:** 339 queries

### With the PR, I empty smarty cache, then go on FO home page:

**First load:** 255 queries
**Second load:** 56 queries

## With thePR, when adding the given piece of code:

`if ($compile_id === null) {
            $compile_id = Context::getContext()->shop->theme->getName();
        }`
at the beginning of `isCached` function in `Module.php`, and emptying smarty cache, I load FO's home page:

**First load:** 663 queries
**Second load:** 56 queries

Let's wait for the CI but it seems good 👍

**Review by matks — APPROVED** [review]

**Comment by kpodemski** [comment]
@PrestaShop/qa-team can we accept @matthieu-rolland tests as QA Approved?

**Review by sowbiba — APPROVED** [review]

**Comment by sarahdib** [comment]
> @PrestaShop/qa-team can we accept @matthieu-rolland tests as QA Approved?

Hello @kpodemski yes you can mege this PR :white_check_mark:

**Comment by kpodemski** [comment]
Thanks @lukaslau @matthieu-rolland and everyone involved.

**Comment by francoispeyret** [comment]
👏 thanks !

**Pull request merged**